starknet-client/src/client/chain_state.rs

```diff
@@ -235,14 +236,16 @@ impl StarknetClient {
 
         // Paginate backwards in block history
         let earliest_block = self.config.earliest_block();
-        'outer: for end_block in (earliest_block..=current_block)
+
+        let mut end_block = BlockId::Tag(BlockTag::Latest);
+        let start_block_limit = current_block.saturating_sub(BLOCK_PAGINATION_WINDOW as u64);
+        'outer: for start_block in (earliest_block..=start_block_limit)
             .rev()
             .step_by(BLOCK_PAGINATION_WINDOW)
         {
             // Exhaust events from the start block to the end block
-            let start_block = end_block.saturating_sub(BLOCK_PAGINATION_WINDOW as u64);
             filter.from_block = Some(BlockId::Number(start_block));
-            filter.to_block = Some(BlockId::Number(end_block));
+            filter.to_block = Some(end_block);
```

Comment:
on the first iteration of this loop, we may fetch more than `BLOCK_PAGINATION_WINDOW` blocks worth of events. we have `from_block = current_block - BLOCK_PAGINATION_WINDOW` and `end_block = LATEST_BLOCK`. is this expected? feels like we should be initializing `end_block = current_block`?

Reply:
Yes this is expected, we don't particularly care if we're off by a block, recall that we paginate within this window anyways.